wasm2c/wasm-rt.h

```diff
@@ -60,52 +60,165 @@ extern "C" {
 #endif
 
 /**
- * Enable memory checking via a signal handler via the following definition:
+ * Backward compatibility: Convert the previously exposed
+ * WASM_RT_MEMCHECK_SIGNAL_HANDLER macro to the ALLOCATION and CHECK macros that
+ * are now used.
+ */
+#if defined(WASM_RT_MEMCHECK_SIGNAL_HANDLER)
+
+#if WASM_RT_MEMCHECK_SIGNAL_HANDLER
+#define WASM_RT_MEMORY_ALLOCATION_MMAP 1
+#define WASM_RT_MEMORY_CHECK_GUARD 1
+#else
+#define WASM_RT_MEMORY_ALLOCATION_MALLOC 1
+#define WASM_RT_MEMORY_CHECK_BOUNDS 1
+#endif
+
+#warning \
+    "WASM_RT_MEMCHECK_SIGNAL_HANDLER has been deprecated in favor of WASM_RT_MEMORY_ALLOCATION_ and WASM_RT_MEMORY_CHECK_ macros"
+#endif
+
+/**
+ * Set the allocation strategy for Wasm memories.
+ *
+ * MMAP: the memory allocation and growth is performed with mmap and mprotect
+ * (or their Windows equivalents) respectively.
  *
- * #define WASM_RT_MEMCHECK_SIGNAL_HANDLER 1
+ * MALLOC: the memory allocation and growth is performed with malloc and realloc
+ * respectively.
  *
- * This is usually 10%-25% faster, but requires OS-specific support.
+ * This defaults to MMAP on 64-bit platforms (so we can use the GUARD based
+ * range checks below) and MALLOC on 32-bit platforms to minimize memory
+ * consumption.
  */
+#ifndef WASM_RT_MEMORY_ALLOCATION_MMAP
+#define WASM_RT_MEMORY_ALLOCATION_MMAP 0
+#endif
 
-#ifndef WASM_RT_SKIP_SIGNAL_RECOVERY
-#define WASM_RT_SKIP_SIGNAL_RECOVERY 0
+#ifndef WASM_RT_MEMORY_ALLOCATION_MALLOC
+#define WASM_RT_MEMORY_ALLOCATION_MALLOC 0
 #endif
 
-/** Signal handler not supported on 32-bit platforms. */
-#if UINTPTR_MAX > 0xffffffff
+#if WASM_RT_MEMORY_ALLOCATION_MMAP && WASM_RT_MEMORY_ALLOCATION_MALLOC
+#error \
+    "Cannot use both WASM_RT_MEMORY_ALLOCATION_MMAP and WASM_RT_MEMORY_ALLOCATION_MALLOC"
 
-#define WASM_RT_SIGNAL_RECOVERY_SUPPORTED 1
+#elif !WASM_RT_MEMORY_ALLOCATION_MMAP && !WASM_RT_MEMORY_ALLOCATION_MALLOC
 
-/* Signal handler is supported. Use it by default. */
-#ifndef WASM_RT_MEMCHECK_SIGNAL_HANDLER
-#ifdef SUPPORT_MEMORY64
-#define WASM_RT_MEMCHECK_SIGNAL_HANDLER 0
+#if UINTPTR_MAX > 0xffffffff
+#undef WASM_RT_MEMORY_ALLOCATION_MMAP
+#define WASM_RT_MEMORY_ALLOCATION_MMAP 1
 #else
-#define WASM_RT_MEMCHECK_SIGNAL_HANDLER 1
+#undef WASM_RT_MEMORY_ALLOCATION_MALLOC
+#define WASM_RT_MEMORY_ALLOCATION_MALLOC 1
 #endif
+
 #endif
 
+/**
+ * Set the range checking strategy for Wasm memories.
+ *
+ * GUARD:  memory accesses rely on unmapped pages/guard pages to trap
+ * out-of-bound accesses.
+ *
+ * BOUNDS: memory accesses are checked with explicit bounds checks.
+ *
+ * This defaults to GUARD as this is the fasest option, iff the requirements of
+ * GUARD --- 64-bit platforms, MMAP allocation strategy, no 64-bit memories ---
+ * are met. This falls back to BOUNDS otherwise.
+ */
+#ifndef WASM_RT_MEMORY_CHECK_GUARD
+#define WASM_RT_MEMORY_CHECK_GUARD 0
+#endif
+
+#ifndef WASM_RT_MEMORY_CHECK_BOUNDS
+#define WASM_RT_MEMORY_CHECK_BOUNDS 0
```

Comment:
Same here.. since this is basically a binary setting can we just have a single macro?  

Doesn't the existing `WASM_RT_MEMCHECK_SIGNAL_HANDLER` means that same thing as the new `WASM_RT_MEMORY_CHECK_GUARD`?

Reply:
Binary setting - will fix!

> WASM_RT_MEMCHECK_SIGNAL_HANDLER vs WASM_RT_MEMORY_CHECK_GUARD

Perhaps, but part of the problem is that the terminology here is confusing, and hence I am not even sure what the WASM_RT_MEMCHECK_SIGNAL_HANDLER  should really be. WASM_RT_MEMORY_CHECK_GUARD on the other hand is clear --- Semantically, it indicates that OOB's should be **prevented** with guard pages (vs bounds checks). Whether we use a signal handler is more of an internal detail when implementing the behavior needed for `WASM_RT_MEMORY_CHECK_GUARD` and `MustTrap`. By decoupling the check from the handling, this can be much clearer.

Reply:
Re the binary setting on this --- I do intend to add other enforcement mechanisms here that are neither guard pages or bounds checks. To avoid making another breaking change in the future, it may be useful to leave this is a separate macros. 

Alternately, if you think we are better off leaving this as binary for now, and making the change later, that is fine too. Let me know the preferred path.